### ISSUE TITLE
Flysystem s3 update breaks thumbnailing behavior

### ISSUE THREAD

**Issue by Arjen-Smit** [sulu/sulu#5468]
| Q | A
| --- | ---
| Sulu Version | Since 2.0.0
| Browser Version | any

# Actual Behavior

You get the following warning `Warning: mime_content_type(): stream does not support seeking` on https://github.com/sulu/sulu/blob/master/src/Sulu/Bundle/MediaBundle/Media/ImageConverter/MediaImageExtractor.php#L52 when doing any media resize operation. This is caused by https://flysystem.thephpleague.com/v1/docs/adapter/aws-s3-v3/#streamed-reads.

# Expected Behavior

media operations should work with newest version of flysystem-aws-s3-v3 (1.0.28+)

# Steps to Reproduce

Base sulu installation where the media is stored on s3.

**config/packages/sulu_media.yaml**
```
sulu_media:
    storage: s3
    storages:
        s3:
            key: '%env(S3_MEDIA_BUCKET_KEY)%'
            secret: '%env(S3_MEDIA_BUCKET_SECRET)%'
            bucket_name: '%env(S3_MEDIA_BUCKET_NAME)%'
            endpoint: '%env(S3_MEDIA_BUCKET_ENDPOINT)%'
            path_prefix: 'media'
            region: 'eu-west-1'
            arguments:
                use_path_style_endpoint: true
                bucket_endpoint: false
```
after that make sure league/flysystem-aws-s3-v3 is at 1.0.28 and you should get an error on any media resize operation

# Possible Solutions

### solution 1 
For now I locked in the version of flysystem-aws-s3-v3
```
"league/flysystem-aws-s3-v3": ">1 <=1.0.24",
```

### solution 2
Flysystems already made a suggestion that it's possible return to the previous behavior by calling the AwsS3Adapter with an additional false.
```
$adapter = new AwsS3Adapter($client, 'your-bucket-name', 'optional/path/prefix', [], false /** disable streamed reads **/);
```

### solution 3
A more permanent solution might be to read out the wrapper_data of the stream as it already contains the information for which the `mime_content_type` is currently called for.
<img width="439" alt="Screenshot 2020-08-27 at 12 07 01" src="https://user-images.githubusercontent.com/697189/91427231-e03e4000-e85d-11ea-83cc-130565e77881.png">
I'm not sure if this will work for other (cloud) storage systems

**Comment by alexander-schranz** [comment]
Thank you for report, found the connected flysystem aws issue here: https://github.com/thephpleague/flysystem-aws-s3-v3/issues/212

*solution 3*  Would maybe be the best but the problem is that a normal resource will not contain that data. Is it possible to detected the `wrapper_type` and go on the `content-type`  if its http and else fallback to the exist mime_content_type? Do you want to create a pull request for it?

**Comment by Arjen-Smit** [comment]
I would target the 2.x branches as this package and feature was introduced with this #4321 so any 2.x who runs composer update might run into this issue.

Before looking into PR to solve this issue I might need some more feedback towards my suggestions. especially since suggestion 3 is not very agnostic if changed in the `MediaImageExtractor`.

**Comment by alexander-schranz** [comment]
> I would target the 2.x branches as this package and feature was introduced with this #4321 so any 2.x who runs composer update might run into this issue.

#4321 did target 2.0 version so the bugfix should be for the `release/2.0` branch.

> Before looking into PR to solve this issue I might need some more feedback towards my suggestions. especially since suggestion 3 is not very agnostic if changed in the MediaImageExtractor.

Not sure what you mean, as said if the content-type is not available it should fallback to the old behaviour so it should work for all adapters also for the local adapter which is no http wrapper_type.

**Comment by Arjen-Smit** [comment]
> Not sure what you mean, as said if the content-type is not available it should fallback to the old behaviour so it should work for all adapters also for the local adapter which is no http wrapper_type.

In that case I would actually think solution 2 might be better as it doesn't change any of the current behaviour of the MediaBundle and reverts the behaviour of `flysystem-aws-s3-v3` back to pre 1.0.28 behaviour.

**Comment by alexander-schranz** [comment]
Is okay for me at current state. The AwsAdapter instance is created here: https://github.com/sulu/sulu/blob/d7b430ceaabdc3f221ff795124b01fe6db036912/src/Sulu/Bundle/MediaBundle/Resources/config/services_storage_s3.xml#L17-L21 so that would be the part we need to change.

**Comment by nnatter** [comment]
For completeness, this issue produced the following log on our server:
```
[2020-09-28T10:49:53.023158+02:00] request.INFO: Matched route "sulu_media.website.image.proxy". {"route":"sulu_media.website.image.proxy","route_parameters":{"_route":"sulu_media.website.image.proxy","_controller":"sulu_media.media_stream_controller:getImageAction","_requestAnalyzer":false,"slug":"991x250/08/8-marshall.jpg"},"request_uri":"https://sulu-demo/uploads/media/991x250/08/8-marshall.jpg?inline=1&v=1-0","method":"GET"} []
[2020-09-28T10:49:53.333689+02:00] php.WARNING: Warning: mime_content_type(): stream does not support seeking {"exception":"[object] (ErrorException(code: 0): Warning: mime_content_type(): stream does not support seeking at /var/www/html/vendor/sulu/sulu/src/Sulu/Bundle/MediaBundle/Media/ImageConverter/MediaImageExtractor.php:52)"} []
[2020-09-28T10:49:53.333875+02:00] php.WARNING: Warning: mime_content_type(): stream does not support seeking {"exception":"[object] (ErrorException(code: 0): Warning: mime_content_type(): stream does not support seeking at /var/www/html/vendor/sulu/sulu/src/Sulu/Bundle/MediaBundle/Media/ImageConverter/MediaImageExtractor.php:52)"} []
[2020-09-28T10:49:53.337719+02:00] app.ERROR: Could not read image from resource {"exception":"[object] (Sulu\\Bundle\\MediaBundle\\Media\\Exception\\InvalidFileTypeException(code: 5004): Could not read image from resource at /var/www/html/vendor/sulu/sulu/src/Sulu/Bundle/MediaBundle/Media/ImageConverter/ImagineImageConverter.php:162)"} []
```

**Comment by plozmun** [comment]
With Azure i had that problem. Look this https://github.com/sulu/sulu/blob/master/src/Sulu/Bundle/MediaBundle/Media/Storage/AzureBlobStorage.php#L61

Another issue is that adapters does not contains client, it must be injected in constructor. This line will fail https://github.com/sulu/sulu/blob/master/src/Sulu/Bundle/MediaBundle/Media/Storage/S3Storage.php#L45

**Comment by alexander-schranz** [comment]
@plozmun @Arjen-Smit I created a bugfix for 2.0 (#5526) and 2.1 (#5527) would be great if somebody could test S3 and Azure Storage. I currently only have a minio S3 to test, would be great if somebody can confirm that it works as expected with this changes.

**Comment by nnatter** [comment]
Thanks again for reporting this. Should hopefully be fixed by #5526
